File: backend/app/schemas/mindmap.py
"""
Mindmap Pydantic schemas for routes.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Shared
# ---------------------------------------------------------------------------

class MindmapStatusOut(str):
    pass


# ---------------------------------------------------------------------------
# Request bodies
# ---------------------------------------------------------------------------

class GenerateMindmapRequest(BaseModel):
    force: bool = False


class AskInThreadRequest(BaseModel):
    content: str
    active_node_id: Optional[str] = None


# ---------------------------------------------------------------------------
# Response shapes
# ---------------------------------------------------------------------------

class MindmapOut(BaseModel):
    id: int
    file_id: int
    classroom_id: int
    status: str
    root_topic: Optional[str] = None
    root_description: Optional[str] = None
    tree_data: Optional[Any] = None
    node_count: int
    generation_meta: Optional[Any] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ExploreNodeResponse(BaseModel):
    already_explored: bool
    last_message_id: Optional[int] = None
    marker_id: Optional[int] = None
    placeholder_id: Optional[int] = None


class ChatMessageOut(BaseModel):
    id: int
    mindmap_id: int
    user_id: int
    node_id: Optional[str] = None
    role: str
    content: str
    message_metadata: Any
    created_at: datetime

    model_config = {"from_attributes": True}


class ChatHistoryResponse(BaseModel):
    messages: list[ChatMessageOut]
    has_more: bool


class AskResponse(BaseModel):
    message: ChatMessageOut

File: frontend/api/mindmap.ts
/**
 * Mindmap API client.
 */
import api from "./axios";

// ── Types ──────────────────────────────────────────────────────────────────

export type MindmapStatus = "pending" | "generating" | "ready" | "failed";

export interface MindmapNodeData {
  id: string;
  topic: string;
  description: string;
  depth: number;
  has_children?: boolean;
  children: MindmapNodeData[];
}

export interface MindmapTree {
  version: number;
  root: MindmapNodeData;
}

export interface Mindmap {
  id: number;
  file_id: number;
  classroom_id: number;
  status: MindmapStatus;
  root_topic: string | null;
  root_description: string | null;
  tree_data: MindmapTree | null;
  node_count: number;
  generation_meta: Record<string, unknown> | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface GenerateMindmapResponse {
  detail: string;
  mindmap: Mindmap;
}

export interface ExploreNodeResponse {
  already_explored: boolean;
  last_message_id: number | null;
  marker_id: number | null;
  placeholder_id: number | null;
}

export type ChatMessageRole = "user" | "assistant" | "marker";

export interface ChatMessage {
  id: number;
  mindmap_id: number;
  user_id: number;
  node_id: string | null;
  role: ChatMessageRole;
  content: string;
  message_metadata: Record<string, unknown>;
  created_at: string;
}

export interface ChatHistoryResponse {
  messages: ChatMessage[];
  has_more: boolean;
}

// ── API calls ──────────────────────────────────────────────────────────────

export const mindmapApi = {
  /** Trigger or resume generation. Returns immediately with current status. */
  generate: (fileId: number, force = false): Promise<GenerateMindmapResponse> =>
    api.post(`/mindmap/files/${fileId}/generate`, { force }).then((r) => r.data),

  /** Poll for current mindmap state (and tree_data once READY). */
  get: (fileId: number): Promise<Mindmap> =>
    api.get(`/mindmap/files/${fileId}`).then((r) => r.data),

  /** Delete the mindmap tree (owner only). */
  delete: (fileId: number): Promise<void> =>
    api.delete(`/mindmap/files/${fileId}`).then(() => undefined),

  /** Start node exploration — returns existing summary or kicks off generation. */
  exploreNode: (mindmapId: number, nodeId: string): Promise<ExploreNodeResponse> =>
    api
      .post(`/mindmap/${mindmapId}/nodes/${nodeId}/explore`)
      .then((r) => r.data),

  /** Fetch chat history. */
  getChatHistory: (mindmapId: number, limit = 50): Promise<ChatHistoryResponse> =>
    api
      .get(`/mindmap/${mindmapId}/chat`, { params: { limit } })
      .then((r) => r.data),

  /** Ask a follow-up question. */
  ask: (
    mindmapId: number,
    content: string,
    activeNodeId: string | null
  ): Promise<{ message: ChatMessage }> =>
    api
      .post(`/mindmap/${mindmapId}/chat/ask`, {
        content,
        active_node_id: activeNodeId,
      })
      .then((r) => r.data),

  /** Clear the user's chat history. */
  clearChat: (mindmapId: number): Promise<void> =>
    api.delete(`/mindmap/${mindmapId}/chat`).then(() => undefined),
};
